File: 152-maximum-product-subarray/152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const maxProduct = function(nums) {
    let result = Math.max(...nums),
        curMin = 1,
        curMax = 1;
    for (const n of nums) {
        // so everything isn't a 0 
        if (n === 0) {
            curMin = 1, curMax = 1;
            continue;
        }
        const temp = curMax*n;
        // important to include n 
        // for cases like [-1, 8]
        curMax = Math.max(curMin*n, temp, n);
        curMin = Math.min(curMin*n, temp, n);
        result = Math.max(result, curMax);
    }
    return result;
};

File: medium/152-maximum-product-subarray/152-maximum-product-subarray.rb
# @param {Integer[]} nums
# @return {Integer}
def max_product(nums)
    res = nums.max
    cur_min = 1
    cur_max = 1
    
    nums.each do |n|
        if n == 0
            cur_min = cur_max = 1
            next
        end
        
        temp = cur_max * n
        cur_max = [cur_min * n, temp, n].max
        cur_min = [cur_min * n, temp, n].min
        res = [res, cur_max].max
    end
    
    res
end